examples/webserver/ratelimit/src/main/java/io/helidon/examples/webserver/ratelimit/Main.java

```diff
@@ -0,0 +1,116 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.helidon.examples.webserver.ratelimit;
+
+import java.util.concurrent.Semaphore;
+
+import io.helidon.config.Config;
+import io.helidon.logging.common.LogConfig;
+import io.helidon.webserver.WebServer;
+import io.helidon.webserver.WebServerConfig;
+import io.helidon.webserver.http.HttpRouting;
+import io.helidon.webserver.http.ServerRequest;
+import io.helidon.webserver.http.ServerResponse;
+
+/**
+ * The application main class.
+ */
+public class Main {
+
+    private static final System.Logger LOGGER = System.getLogger(Main.class.getName());
+
+    private static Semaphore rateLimitSem = null;
+
+    /**
+     * Cannot be instantiated.
+     */
+    private Main() {
+    }
+
+    /**
+     * Application main entry point.
+     * @param args command line arguments.
+     */
+    public static void main(String[] args) {
+
+        // load logging configuration
+        LogConfig.configureRuntime();
+
+        // initialize global config from default configuration
+        Config config = Config.create();
+        Config.global(config);
+
+        WebServerConfig webserverConfig = WebServer.builder()
+                .config(config.get("server"))
+                .routing(Main::routing)
+                .buildPrototype();
+
+        LOGGER.log(System.Logger.Level.INFO, "WebServer maxConcurrentRequests is " + webserverConfig.maxConcurrentRequests());
+
+        WebServer webserver = webserverConfig.build().start();
+
+        System.out.println("WEB server is up! http://localhost:" + webserver.port() + "/sleep");
+    }
+
+    /**
+     * Updates HTTP Routing.
+     */
+    static void routing(HttpRouting.Builder routing) {
+        routing.get("/sleep/{seconds}", Main::sleepHandler);
+        int rateLimit = Config.global().get("app").get("ratelimit").asInt().orElse(20);
+        LOGGER.log(System.Logger.Level.INFO, "         Application rate limit is " + rateLimit);
+        rateLimitSem = new Semaphore(rateLimit);
+    }
+
+    /**
+     * Sleep for a specified number of seconds.
+     * The optional path parameter controls the number of seconds to sleep. Defaults to 1
+     *
+     * @param request  server request
+     * @param response server response
+     */
+    private static void sleepHandler(ServerRequest request, ServerResponse response) {
+        int seconds = request.path().pathParameters().first("seconds").asInt().orElse(1);
+
+        try {
+            rateLimitSem.acquire();
```

Comment:
This looks fine, but perhaps a more elegant solution can be crafted using FT bulkheads, which are designed to limit concurrency. Not sure if the resulting code would be more readable, but may be worth a try.

Reply:
Good point. I liked showing the basics of using the Semaphore -- since it is simple.  I will do another version using bulkheads and see how it tastes.

Reply:
I've added a `sleep-bulkhead` endpoint to demonstrate using Bulkhead